roqoqo-iqm/src/backend.rs

```diff
@@ -359,42 +294,48 @@ impl Backend {
     /// # Arguments
     ///
     /// * `circuit` - The [roqoqo::Circuit] to be checked
-    pub fn validate_circuit(&self, circuit: &Circuit) -> Result<(), RoqoqoBackendError> {
+    pub fn validate_circuit(&self, circuit: &Circuit) -> Result<(), IqmBackendError> {
         // Check that the circuit doesn't contain more qubits than the device supports
         let mut measured_qubits: Vec<usize> = vec![];
-        let number_qubits = match _get_number_qubits(circuit) {
-            Some(x) => x,
-            None => {
-                return Err(RoqoqoBackendError::GenericError {
-                    msg: "Empty circuit was passed to the backend.".to_string(),
-                })
+        let number_qubits = _get_number_qubits(circuit).ok_or(IqmBackendError::EmptyCircuit)?;
+
+        // NOTE checking also the name is a workaround for a pyo3 deserialization bug that causes
+        // the if let to match even when the device is not Deneb. This issue should have been fixed
+        // by removing the bincode deserialization attempt in the device pyo3 files, but I leave the
+        // name check here for extra insurance.
+        if self.device.name() == "Deneb" {
+            if let IqmDevice::DenebDevice(device) = &self.device {
+                device.validate_circuit(circuit)?
             }
-        };
-
-        if let IqmDevice::DenebDevice(device) = &self.device {
-            device.validate_circuit(circuit)?
         } else {
-            self.validate_circuit_connectivity(circuit)?
+            self.validate_circuit_connectivity(circuit).map_err(|err| {
+                IqmBackendError::InvalidCircuit {
+                    msg: err.to_string(),
+                }
+            })?
         }
 
         // Check that
-        // 1) Every qubit is only measured once
+        // 1) Every qubit is measured exactly once
         // 2) Output registers are large enough
+        let mut measured = false;
         for op in circuit.iter() {
             match op {
                 Operation::MeasureQubit(o) => {
+                    measured = true;
                     let qubit = *o.qubit();
                     if measured_qubits.contains(&qubit) {
-                        return Err(RoqoqoBackendError::GenericError {
-                            msg: format!("Qubit {} is being measured more than once.", &qubit),
+                        return Err(IqmBackendError::InvalidCircuit {
+                            msg: format!("Qubit {} is being measured multiple times.", qubit),
                         });
                     } else {
                         measured_qubits.push(qubit)
                     }
                 }
                 Operation::PragmaRepeatedMeasurement(o) => {
+                    measured = true;
                     if !measured_qubits.is_empty() {
-                        return Err(RoqoqoBackendError::GenericError {
+                        return Err(IqmBackendError::InvalidCircuit {
```

Comment:
I feel like I've reviewed a lot of these changes before, is your branch up to date with upstream/main?